### PR TITLE
feat: create initial working datadog metric extractor

### LINKED ISSUE #1: Create initial implementation
Create initial implementation triggered by static event to pull a set of datadog metrics and save them to S3.

### PR DESCRIPTION
**Pull request by bbangert**
Initial serverless project template with deployable lambda that
can be triggered off scheduled Cloudwatch events. Contains README
on usage.

Closes #1

### CODE REVIEW

File: ddsink/handler.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    format_metric_results,
+    MetricConfig,
+)
+from ddsink.types import JSONDict
+
+logger = logging.getLogger()
+logger.setLevel(logging.INFO)
+
+
+def save_output_to_s3(config: MetricConfig, output: Iterator[JSONDict])-> None:
```
Comment (pjenvey) [inline ddsink/handler.py]: should be whitespace in between ) and ->

File: ddsink/handler.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    dt = datetime.datetime.fromtimestamp(config.end_time)
+    file_name = config.s3_path.format(
```
Comment (pjenvey) [inline ddsink/handler.py]: Totally unnecessary here but FYI: format strings can drive all the sub formatting from themselves now, e.g.:

'date={dt:%Y-%m-%d}/hour={dt:%H}'.format(dt=dt)

File: ddsink/handler.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    ))
+    data = gzip.compress(b"".join(json.dumps(entry).encode('utf-8') for entry
+                                  in output))
+    s3 = boto3.resource('s3')
+    object = s3.Object(config.s3_bucket, file_name)
+    object.put(Body=data, ContentEncoding="gzip")
+
+
+def lambda_handler(event: JSONDict, context=None):
+    logger.info("Running with event: %s",event)
```
Comment (pjenvey) [inline ddsink/handler.py]: ," whitespace

File: ddsink/handler.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+        filename=file_name,
+    ))
+    data = gzip.compress(b"".join(json.dumps(entry).encode('utf-8') for entry
+                                  in output))
+    s3 = boto3.resource('s3')
+    object = s3.Object(config.s3_bucket, file_name)
+    object.put(Body=data, ContentEncoding="gzip")
+
+
+def lambda_handler(event: JSONDict, context=None):
```
Comment (pjenvey) [inline ddsink/handler.py:44]: nitpick context not needed here.. (currently anyway)

File: ddsink/metrics.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+        if not val:
+            continue
+        ts = datetime.datetime.fromtimestamp(epoch_time/1000.0).strftime(
+            "%Y-%m-%d %H:%M:%S"
+        )
+        yield dict(timestamp=ts, value=val)
+
+
+def extract_scope(scope: str) -> JSONDict:
+    return dict([part.split(':', 1) for part in scope.split(',') if part])
```
Comment (pjenvey) [inline ddsink/metrics.py]: nitpick: list comp could be a gen exp

**Review by pjenvey — DISMISSED** [review]
(fix the whitespace)

File: ddsink/handler.py (hunk 4 above)
Comment (bbangert) [reply ddsink/handler.py:44]: It isn't, but the lambda event could pass it in.

File: ddsink/handler.py [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+    initialize_datadog,
+    load_config,
+    query_datadog,
+    format_metric_results,
+    MetricConfig,
+)
+from ddsink.types import JSONDict
+
+logger = logging.getLogger()
+logger.setLevel(logging.INFO)
```
Comment (jrconlin) [inline ddsink/handler.py:21]: Make an env, incase things ever get too chatty?

**Review by jrconlin — APPROVED** [review]

File: ddsink/handler.py (hunk 6 above)
Comment (bbangert) [reply ddsink/handler.py:21]: It's just logged out to Cloudwatch, so it should be ok.

**Pull request merged**